Enabling Markdown for GristDocTour

ogui11aume opened the pull request:
## Context

So far Grist only permits raw text to appear as a one-liner in modal during a document tour : not a very nice user experience for somebody to get a first impression of your super Grist document. 🙁
With Markdown, the document tours are much better: 
* rich text with new lines and line spacing 
* multiple clickable links (multiple) directly within the text, not handled as a separate data
* images

## Proposed solution

Technically : 
Just parsing the `Body` cell from `GristDocTour` table with already available `marked` library, sanitize and following the renderer approach to better select which aspect of markdown to render.

Documentation : https://github.com/ogui11aume/hackdays2025/blob/ergonogrist/submissions/ergonogrist/assets/markdown-for-gristdoctour/deliverable1.md 

## Related issues

Didn't find any issue ( and I didn't open one although I made the change at the request of other users).

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help I created a sample Document to rapidly test the behaviour here https://github.com/ogui11aume/hackdays2025/blob/ergonogrist/submissions/ergonogrist/assets/markdown-for-gristdoctour/Markdown-for-GristDocTour.grist 

## Screenshots / Screencasts

https://github.com/ogui11aume/hackdays2025/blob/ergonogrist/submissions/ergonogrist/assets/markdown-for-gristdoctour/deliverable1.md 

![Example of Markdown during a document tour](https://github.com/user-attachments/assets/f3918453-8b4c-467d-8f2e-03e84c87d5c1)

ogui11aume committed:
Enabling Markdown for GristDocTour

fflorent reviewed (approved):
Sounds good to me. Thank you very much!

I just wondered why the iframe were allowed in `sanitizeTutorialHTML`. Sounds like it is required and used in the [official tutorial](https://templates.getgrist.com/doc/woXtXUBmiN5TGhoXCBGSBf) to embed a Youtube video.

We will ask for an official review from Grist Labs.

georgegevoian commented on app/client/ui/DocTour.ts:
```diff
@@ -10,6 +10,10 @@ import {IconList, IconName} from 'app/client/ui2018/IconList';
 import sortBy = require('lodash/sortBy');
+import {marked} from "marked";
+import {renderer} from 'app/client/ui/DocTutorialRenderer';
+import {sanitizeTutorialHTML} from "./sanitizeHTML";
```
```suggestion
import {sanitizeHTML} from "app/client/ui/sanitizeHTML";
```

You can use this variant instead, unless you need to tweak the default sanitization behavior, in which case you can just add a new variant like `sanitizeTourHTML`.

If no tweaking is necessary, you can just use the `markdown` helper from `app/client/lib/markdown.ts`.

Minor lint: absolute module path is preferred over relative.

georgegevoian commented on app/client/ui/DocTour.ts:
```diff
@@ -10,6 +10,10 @@ import {IconList, IconName} from 'app/client/ui2018/IconList';
 import {dom} from 'grainjs';
 import sortBy = require('lodash/sortBy');
+import {marked} from "marked";
+import {renderer} from 'app/client/ui/DocTutorialRenderer';
```
```suggestion
import {renderer} from 'app/client/ui/DocTourRenderer';
```

georgegevoian replied:
`DocTourRenderer` isn't being imported here.

georgegevoian commented on app/client/ui/DocTour.ts:
```diff
@@ -44,7 +48,17 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
     }
     const title = getValue("Title");
-    let body: HTMLElement | string = getValue("Body");
+
+    const bodyHtmlContent = sanitizeTutorialHTML(marked.parse(getValue("Body"), {
+      async: false, renderer
+    }));
+    const element = document.createElement('div');
+    element.innerHTML = bodyHtmlContent;
+
+
+    let body: HTMLElement = element;
+
+
```
I think you can replace the `dom('p', body),` below to something like `dom('p', markdown(body)),` or `dom('p', el => el.innerHTML = bodyHtmlContent)`, depending on whether you're using the `markdown` helper or your own renderer.

georgegevoian commented:
Tests are now in grist-core: https://github.com/gristlabs/grist-core/commit/2997ca8c9688dba6743d6387b2cb1adb80c91b43.

berhalak commented on app/client/ui/DocTour.ts:
```diff
@@ -44,7 +48,17 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
     const title = getValue("Title");
-    let body: HTMLElement | string = getValue("Body");
+
+    const bodyHtmlContent = sanitizeTutorialHTML(marked.parse(getValue("Body"), {
```
Can you use `sanitizeHTMLIntoDOM` instead? We should avoid setting `innerHTML` directly?

berhalak commented on app/client/ui/DocTourRenderer.ts:
```diff
@@ -0,0 +1,8 @@
+
+export const renderer = new marked.Renderer();
+
+renderer.link = ({href, text}) =>
```
I think that the default `marked.Renderer` allows arbitrary html, which potentially allows some phishing attacks. For example with `<div style="position: fixed; color: red; inset: 0">Boo</div>` you can cover whole Grist layout and provide your own.

What do you think about reusing the renderer from `app/client/ui/MarkdownCellRenderer.ts`? It clear out any html content.

ogui11aume replied:
I used ALLOWED_TAGS with marked.Renderer in app/client/ui/sanitizeHTML.ts to limit abuse and only allow those HTML elements that have been tested (+ table).  
- Agreed, this does not yet avoid the problem you described ( covering div ). To avoid the problem in your example, I would also need to limit the usage of style at least in div, should I go that far ?
- The problem with app/client/ui/MarkdownCellRenderer.ts is that, according to comment, it will not permit images or other HTML elements. So I didn't go that way.
https://github.com/gristlabs/grist-core/blob/129a053af02ee7e08fe58ebe1f9451e0f2d3f22e/app/client/ui/MarkdownCellRenderer.ts#L59 : 
>  This does not support HTML or images in the markdown value, and renders those as text.

georgegevoian replied:
@ogui11aume, what if we disallowed HTML for the time being?

```javascript
renderer.html = ({raw}) => escape(raw);
```

Would there be an immediate need for tours to support arbitrary HTML?

Having a robust and shared sanitization approach across different UI components like tutorials, tours, etc. seems desirable, but could perhaps be reconciled later. (Tutorials in particular should probably also disallow HTML outside of the specific iframe that's permitted, but that's unrelated to your proposed changes.)

ogui11aume commented:
I have read the CLA Document and I hereby sign the CLA

ogui11aume committed:
Tests for Markdown in DocTour

ogui11aume commented:
> @ogui11aume Do you need our review again?
Yes please, I need a new review. 
For my commits I have : 
* Added the mail I'm signing my commits with in Github account and verified it. 
* Signed my commits 
* Taken into account your advice regarding some code refactoring (thanks @georgegevoian, I've only just seen @berhalak `sanitizeHTMLIntoDOM` suggestion from last week )
* Fixed existing nbrowser tests without altering content in existing lines in doctour.grist (thanks @hexaltation )
* Added new lines in doctour.grist to test rendering Markdown formatting syntax, headings, image, link, quote, lists (ordered and unordered), code blocks. 

Most other Markdown extended syntax are not currently being rendered so I didn't create tests for them. Let me know if you thinki other tests would be useful.

ogui11aume commented on test/nbrowser/DocTour.ts:
```diff
@@ -154,7 +265,9 @@ describe('DocTour', function () {
   async function checkDocTourPresent() {
     // Check the expected message.
-    assert.match(await driver.findWait('.test-onboarding-popup', 500).getText(), /General Kenobi/);
+     assert.isFalse(await driver.findWait('.test-onboarding-popup', 500)
```
Hi
I agree : this test should assert _True_ that the popup for DocTour shows up since the intended purpose of the test is : 
 > should show doctour for a new anon user, error-free, without welcome tour

I fix it to true, Thanks @georgegevoian .

ogui11aume committed:
fix(test): Change the parent html element and change assert to true

Discussion regarding the changes
https://github.com/gristlabs/grist-core/pull/1653#discussion_r2249520614

ogui11aume committed:
Refactor DOM_FRAGMENT and ALLOWED_TAGS

Refactor using DOM_FRAGMENT like in sanitizeHTMLIntoDOM
https://github.com/gristlabs/grist-core/pull/1653#discussion_r2231107030
and specify allowed HTML tags to limit abuse according to
https://github.com/gristlabs/grist-core/pull/1653#discussion_r2231117948

fflorent commented:
@ogui11aume The LeftPanel test is unstable, not related to your work (though I can't make it fail locally).
I take a look a the tests for the Grist Tour.

fflorent reviewed (commented):
@ogui11aume I don't seem to be able to make the test file fail locally (I ran the `DocTour` tests 20 times with no failure).

Do you reproduce the failure on your side?

fflorent commented:
@ogui11aume I was trying to reproduce this failure your [mentioned earlier](https://github.com/gristlabs/grist-core/pull/1653#issuecomment-3144729777):
https://github.com/gristlabs/grist-core/actions/runs/16676842047/job/47205430863?pr=1653#step:23:850

As Github collapses the resolved comments, I miss that point (I should be more vigilant next time).

Anyway, glad you fixed the issue :)

georgegevoian commented on app/client/ui/DocTour.ts:
```diff
@@ -44,17 +47,22 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
+    if (!title && !(bodyValue.trim()) ) {
+      return null;
+    }
+    //const bodyHtmlContent
```
```suggestion
```

dsagal commented:
What do you think of the suggestion in #1864 to respect the setting of the `Body` field in the `GristDocTour` table, and only rendering markdown when that field is set to markdown? The main reason I am thinking of it is to avoid accidentally affecting existing doc-tours.

If we do that, it would also make sense to me to reuse the same markdown renderer + sanitizing as used for cells. You would then be able to expect the doc tour to show content in the same way you see it in the cells of the `GristDocTour` table. What do you think?

ogui11aume committed:
feat(doc-tour): disable HTML rendering in DocTour Markdown
- Prevent raw HTML from being rendered within DocTour markdown content
- Improves security and consistency of displayed documentation
- Updated related tests accordingly

ogui11aume committed:
feat(doc-tour): disable IMG rendering in DocTour Markdown using MarkdownCellRenderer

dsagal commented:
> I haven't found how to test whether the column **Cell Format** is **Markdown**. Any idea on how I should go to do that ?

Oh, yes. The column info isn't readily exposed in this case, so it takes some digging through metadata to extract it. In other words, excellent question :)

This should work:
```
diff --git a/app/client/ui/DocTour.ts b/app/client/ui/DocTour.ts
index d1c2385f5..61aa2237d 100644
--- a/app/client/ui/DocTour.ts
+++ b/app/client/ui/DocTour.ts
@@ -8,6 +8,7 @@ import {IOnBoardingMsg, startOnBoarding} from 'app/client/ui/OnBoardingPopups';
 import {isNarrowScreen} from 'app/client/ui2018/cssVars';
 import {IconList, IconName} from 'app/client/ui2018/IconList';
 import {DocData} from 'app/common/DocData';
+import {safeJsonParse} from 'app/common/gutil';
 import {dom} from 'grainjs';
 import sortBy = require('lodash/sortBy');

@@ -39,6 +40,10 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
   await docData.fetchTable(tableId);
   const tableData = docData.getTable(tableId)!;

+  const tableRef = docData.getMetaTable("_grist_Tables").findRow('tableId', tableId);
+  const bodyCol = docData.getMetaTable("_grist_Tables_column").filterRecords({parentId: tableRef, colId: 'Body'})[0];
+  const widgetType: string|undefined = safeJsonParse(bodyCol?.widgetOptions, {})?.widget;
+  console.warn("BODYCOL", widgetType);
   const result = sortBy(tableData.getRowIds(), tableData.getRowPropFunc('manualSort') as any).map(rowId => {
     function getValue(colId: string): string {
       return String(tableData.getValue(rowId, colId) || "");
```

ogui11aume committed:
feat(doc-tour): Make DocTour markdown rendering conditional based on column Body Cell Format specified as Markdown

ogui11aume committed:
feat(doc-tour): Test file with Body column Cell Format 'markdown' and explicit titles

dsagal commented on app/client/ui/DocTour.ts:
```diff
@@ -69,8 +92,8 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
       body = dom(
-        'div',
-        dom('p', body),
+        'span',
```
For plain-text, this used to be `<div><p>body</p>...</div>`, but now is `<span>body...</span>`. Is this change intentional?

dsagal commented on app/client/ui/DocTour.ts:
```diff
@@ -43,18 +47,37 @@ async function makeDocTour(docData: DocData, docComm: DocComm): Promise<IOnBoard
+    // Renders Markdown only if the `Body` colum of type `Text` specifies "Cell Format" as `Markdown`
+    const cellFormat = getCellFormat("Body");
+    if(cellFormat == "Markdown") {
+      const element = sanitizeTourHTML(marked.parse(bodyValue, {
```
I suggest importing `renderCellMarkdown` from 'app/client/ui/MarkdownCellRenderer', and using `body = renderCellMarkdown(bodyValue)`. Or if that's not suitable, you could explain why (in a code comment ideally).

If it works, then we don't need a separate sanitizer. I'd rather avoid that because then we have to carefully analyze what sanitization settings are different and if they are all correct.

ogui11aume replied:
Thank you for all the confirmation and explanations.
Moving onward with the choice of `renderCellMarkdown` for DocTour, I'm pushing the changes with updated tests.

dsagal commented on test/nbrowser/DocTour.ts:
```diff
@@ -47,51 +47,175 @@ describe('DocTour', function () {
+      },
+      {
+        body: '<span><p>' +
+          '<strong>bold</strong>' +
```
Nice! I think there may be a reasonably simple way to include a test of plaintext without a new fixture by switching the column type using API from the test (using `PATCH /api/docs/{docId}/tables/GristDocTour/columns` endpoint) in a test case and running some more checks. But also not so important and OK to omit.

The important thing is to test manually that both plaintext and markdown look as expected (e.g. that this doesn't change behavior of plaintext in some unintended way).

ogui11aume committed:
feat(doc-tour): Using MarkdownCellRenderer to render DocTour Body with Markdown format

dsagal reviewed (commented):
Thank you! Looks good, just a couple very small requests.

ogui11aume committed:
feat(doc-tour): Refactoring columnCellFormat and test linting fixes

dsagal reviewed (commented):
Sorry to be slow. I have one more question (and one white space nitpick).
